### PR TITLE
Quest System Part 1

### PR DESCRIPTION
**Pull request by BenCheung0422**
This is a quest system of 2.2.0.
System Base of #203 
Part 1

### CODE REVIEW

**Comment by BenCheung0422** [comment]
A quick note

Survival mode: Quest optional (tutorials only), Recipe unlocking disabled (enable in tutorials)
Story mode: Quest enabled (tutorials optional), Recipe unlocking enable
(and adventure mode?)

File: src/main/java/minicraft/screen/QuestsDisplay.java [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+		reinitializeQuests();
+		for (String n : unlocked) {
+			Quest q = getQuest(n);
+			if (!unlockedQuests.contains(q)) unlockedQuests.add(q);
+		}
+		for (String n : done) doneQuests.add(getQuest(n));
+		for (Entry<String, String> e : data.entrySet()) questStatus.put(e.getKey(), QuestStatus.fromStringType(e.getValue()));
+	}
+	public static void loadGameQuestsObj(ArrayList<Quest> unlocked, ArrayList<Quest> done) { loadGameQuestsObj(unlocked, done, new HashMap<>()); }
+	public static void loadGameQuestsObj(ArrayList<Quest> unlocked, ArrayList<Quest> done, Map<String, String> data) {
```
Comment (Makkkkus) [inline src/main/java/minicraft/screen/QuestsDisplay.java:223]: Why are quests and quest objects loaded separately?

**Review by Makkkkus — CHANGES_REQUESTED** [review]
In the quests display, unlocked and done should be the only available options, and the others should be removed. This is to remove clutter in the menu, and to make it simpler.

Try to get rid of single-line statements, like `if () someFunctionHere();`. These make the code less readable, and look very ugly. Remeber to add proper whitespace!

**Comment by BenCheung0422** [comment]
You can merge this first as the main part of the quest system has been done.

**Comment by Makkkkus** [comment]
I can't see that you have fully followed my requirements. Look over my comments again.

File: src/main/java/minicraft/saveload/Load.java [hunk 2]
```diff
@@ -492,6 +495,40 @@ private void loadWorld(String filename) {
 				}
 			}
 		}
+		if (new File(location+"Quests.json").exists()) {
```
Comment (Makkkkus) [inline src/main/java/minicraft/saveload/Load.java:496]: Formatting

File: src/main/java/minicraft/saveload/Load.java [hunk 3]
```diff
@@ -492,6 +495,40 @@ private void loadWorld(String filename) {
...
+			try {
+				JSONObject questsObj = new JSONObject(loadFromFile(location+"Quests.json", true));
```
Comment (Makkkkus) [inline src/main/java/minicraft/saveload/Load.java:498]: Formatting

File: src/main/java/minicraft/screen/CraftingDisplay.java [hunk 4]
```diff
@@ -124,6 +124,9 @@ public void tick(InputHandler input) {
 						selectedRecipe.getProduct().equals(Items.get("reg clothes"))) {
 					AchievementsDisplay.setAchievement("minicraft.achievement.clothes", true);
 				}
+				if (((ToolItem)selectedRecipe.getProduct()).type.equals(ToolType.Sword)) {
```
Comment (Makkkkus) [inline src/main/java/minicraft/screen/CraftingDisplay.java:127]: Really, add newlines after closing brackets in situations like this.

File: src/main/java/minicraft/saveload/Save.java [hunk 5]
```diff
@@ -231,6 +234,33 @@ private void writeWorld(String filename) {
...
+
+		for (Quest q : QuestsDisplay.getUnlockedQuests()) {
+			unlockedQuests.put(q.id);
+		}
+		for (Quest q : QuestsDisplay.getCompleteQuest()) {
+			doneQuests.put(q.id);
+		}
+		for (Entry<String, QuestsDisplay.QuestStatus> e : QuestsDisplay.getStatusQuests().entrySet()) {
+			questData.put(e.getKey(), e.getValue().toQuestString());
+		}
```
Comment (Makkkkus) [inline src/main/java/minicraft/saveload/Save.java:250]: Add newlines after the closing brackets.

File: src/main/java/minicraft/saveload/Load.java [hunk 6]
```diff
@@ -492,6 +495,40 @@ private void loadWorld(String filename) {
...
+
+				for (int i = 0; i<unlockedQuests.length(); i++) {
+					unlocked.add(unlockedQuests.getString(i));
+				}
+				for (int i = 0; i<doneQuests.length(); i++) {
+					done.add(doneQuests.getString(i));
+				}
+				for (String i : questData.keySet()) {
+					questStatus.put(i, questData.getString(i));
+				}
```
Comment (Makkkkus) [inline src/main/java/minicraft/saveload/Load.java:516]: Add newlines after the closing brackets.

File: src/main/java/minicraft/screen/QuestsDisplay.java [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+				}
+
+				if (unlocked)
+					initiallyUnlocked.add(id);
+
+				quests.put(id, unlocked? new Quest(id, obj.getString("desc"), unlocked, unlocks): new Quest(id, obj.getString("desc"), unlocks));
+			}
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
```
Comment (Makkkkus) [inline src/main/java/minicraft/screen/QuestsDisplay.java:59]: The formatting here is nearly perfect. Everything is perfect other than some operators not having space on both sides.

File: src/main/java/minicraft/screen/QuestsDisplay.java [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+		super.tick(input);
+
+		if (input.getKey("cursor-left").clicked) if (selectedEntry > 0) {
+			selectedEntry--;
+			updateEntries();
+		};
+		if (input.getKey("cursor-right").clicked) if (selectedEntry < 1) {
+			selectedEntry++;
+			updateEntries();
+		};
```
Comment (Makkkkus) [inline src/main/java/minicraft/screen/QuestsDisplay.java:226]: Closing brackets don't need semicolons.

Add newlines after the closing brackets.

File: src/main/java/minicraft/screen/QuestsDisplay.java [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+
+		super.render(screen);
+	}
+
+	private void updateEntries() {
+		menus[0].setEntries(questEntries[selectedEntry]);
+
+		String[] entryNames = new String[] {
+			"Unlocked", "Done"
+		};
```
Comment (Makkkkus) [inline src/main/java/minicraft/screen/QuestsDisplay.java:252]: Another semicolon

**Review by Makkkkus — CHANGES_REQUESTED** [review]

File: src/main/java/minicraft/screen/QuestsDisplay.java (hunk 9 above)
Comment (Makkkkus) [reply src/main/java/minicraft/screen/QuestsDisplay.java:254]: Ah, yeah, that is true. My bad

File: src/main/java/minicraft/core/World.java [hunk 10]
```diff
@@ -126,6 +129,7 @@ public static void resetGame(boolean keepPlayer) {
 			Updater.pastDay1 = false;
 			player.findStartPos(level, WorldGenDisplay.getSeed()); // Finds the start level for the player
 			level.add(player);
+			QuestsDisplay.loadGameQuests(new ArrayList<>(), new ArrayList<>());
```
[the remark this answers is not in the document]
Comment (Makkkkus) [reply src/main/java/minicraft/core/World.java:132]: You still haven't fixed this. Use the resetGameQuests method!

File: src/main/java/minicraft/screen/QuestsDisplay.java (hunk 7 above)
Comment (Makkkkus) [reply src/main/java/minicraft/screen/QuestsDisplay.java:59]: You still haven't fixed the changes I requested.

Comment (BenCheung0422) [reply src/main/java/minicraft/screen/QuestsDisplay.java:60]: Ok

File: src/main/java/minicraft/core/World.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (BenCheung0422) [reply src/main/java/minicraft/core/World.java]: Done

**Review by Makkkkus — APPROVED** [review]

**Pull request merged**